modules/images/hero-image-early-hints/load.php

```diff
@@ -0,0 +1,174 @@
+}
+add_action( 'init', 'perflab_hieh_add_hooks' );
+
+/**
+ * Checks the request URI and based on it attempts to send a 103 Early Hints header for the hero image.
+ *
+ * @since n.e.x.t
+ */
+function perflab_hieh_send_early_hints_header() {
+	global $perflab_hieh_request_uri;
+
+	// Bail if not a frontend request.
+	if ( is_admin() || defined( 'XMLRPC_REQUEST' ) || defined( 'REST_REQUEST' ) || defined( 'MS_FILES_REQUEST' ) ) {
+		return;
+	}
+
+	$perflab_hieh_request_uri = $_SERVER['REQUEST_URI'];
+
+	$home_path = parse_url( home_url(), PHP_URL_PATH );
+	if ( is_string( $home_path ) && '' !== $home_path ) {
+		$home_path       = trim( $home_path, '/' );
+		$home_path_regex = sprintf( '|^%s|i', preg_quote( $home_path, '|' ) );
+
+		$perflab_hieh_request_uri = preg_replace( $home_path_regex, '', $perflab_hieh_request_uri );
+		$perflab_hieh_request_uri = trim( $perflab_hieh_request_uri, '/' );
+	}
+
+	if ( empty( $perflab_hieh_request_uri ) ) {
+		$perflab_hieh_request_uri = '/';
+	}
+
+	$hero_img_srcset = get_option( 'perflab_hieh_' . md5( $perflab_hieh_request_uri ) );
+	if ( ! $hero_img_srcset ) {
+		return;
+	}
+
+	$hero_img_srcset = array_filter(
+		array_map(
+			function( $srcset_entry ) {
+				if ( ! preg_match( '/ (\d+)w$/', $srcset_entry, $matches ) ) {
+					return false;
+				}
+				return array(
+					'src' => str_replace( $matches[0], '', $srcset_entry ),
+					'w'   => (int) $matches[1],
+				);
+			},
+			explode( ', ', $hero_img_srcset )
+		)
+	);
+	$hero_img_srcset = wp_list_sort( $hero_img_srcset, 'w', 'ASC' );
+
+	// This approach is obviously not reliable and clearly shows how Early Hints as of today
+	// is not really feasible for images due to lack of srcset and sizes support.
+	$min_width    = wp_is_mobile() ? 1000 : 1600;
+	$hero_img_url = '';
+	foreach ( $hero_img_srcset as $srcset_entry ) {
+		if ( $srcset_entry['w'] > $min_width ) {
+			$hero_img_url = $srcset_entry['src'];
+			break;
+		}
+	}
+	if ( ! $hero_img_url ) {
+		$hero_img_url = array_pop( $hero_img_srcset )['src'];
+	}
+
+	status_header( 103 );
+	header( "Link: <{$hero_img_url}>; rel=preload; as=image", false );
+
+	// Fix WP core headers no longer being output because of its problematic `headers_sent()` checks.
+	add_filter(
+		'wp_headers',
+		function( $headers ) {
+			// Send headers on 'send_headers' early, since status header will still be sent by WP.
+			add_action(
+				'send_headers',
+				function() use ( $headers ) {
+					if ( isset( $headers['Last-Modified'] ) && false === $headers['Last-Modified'] ) {
+						unset( $headers['Last-Modified'] );
+
+						header_remove( 'Last-Modified' );
+					}
+
+					foreach ( (array) $headers as $name => $field_value ) {
+						header( "{$name}: {$field_value}" );
+					}
+				},
+				// phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+				PHP_INT_MIN
+			);
+			return $headers;
+		}
+	);
+}
+add_action( 'plugins_loaded', 'perflab_hieh_send_early_hints_header' );
```

Comment:
Do we really need all this logic if we can just check the current post thumbnail ID and preload that image if it is available? If there is no thumbnail for the post, then I think we should preload nothing because the first image in the content is not always the hero image. Furthermore, the first image in the content can be far below the fold and in this case preloading it is inefficient. 🤔

Reply:
@eugene-manuilov What you're saying about the first content image can certainly happen, though at the same time it even applies to the featured image, it entirely depends on the theme where that is displayed. It _usually_ is in the header area but not always.

Each of these concerns already apply to the `fetchpriority="high"` and `loading="lazy"` features too to be fair.